File: app/src/main/java/me/shreyasr/chatse/chat/service/ChatWebSocketListener.java
package me.shreyasr.chatse.chat.service;

import android.util.Log;

import com.squareup.okhttp.Response;
import com.squareup.okhttp.ws.WebSocket;
import com.squareup.okhttp.ws.WebSocketListener;

import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;

import me.shreyasr.chatse.util.Logger;
import okio.Buffer;
import okio.BufferedSource;

public class ChatWebSocketListener implements WebSocketListener {

    private final ObjectMapper mapper = new ObjectMapper();
    private final ServiceWebsocketListener listener;
    private final String site;

    public ChatWebSocketListener(String site, ServiceWebsocketListener listener) {
        this.listener = listener;
        this.site = site;
    }

    @Override
    public void onOpen(WebSocket webSocket, Response response) {
        Logger.event(this.getClass(), "websocket open: " + site);
        listener.onConnect(site, true);
    }

    @Override
    public void onFailure(IOException e, Response response) {
        Logger.event(this.getClass(), "websocket fail: " + site);
        listener.onConnect(site, false);
        Log.e(e.getClass().getSimpleName(), e.getMessage(), e);
    }

    @Override
    public void onMessage(BufferedSource payload,
                          WebSocket.PayloadType type) throws IOException {
        String message = payload.readUtf8();
        payload.close();
        Logger.event(this.getClass(), "websocket message: " + site + ": " + message);
        try {
            JsonNode root = mapper.readTree(message);
            listener.onNewEvents(site, root);
        } catch (IOException e) {
            Log.e(e.getClass().getSimpleName(), e.getMessage(), e);
        }
    }

    @Override
    public void onPong(Buffer payload) {
        Logger.event(this.getClass(), "websocket pong: " + site);
    }

    @Override
    public void onClose(int code, String reason) {
        Logger.event(this.getClass(), "websocket close: " + site + ": " + code + ", " + reason);
    }

    interface ServiceWebsocketListener {

        void onNewEvents(String site, JsonNode root);

        void onConnect(String site, boolean success);
    }
}

File: app/src/main/java/me/shreyasr/chatse/chat/service/ChatWebSocketListener.kt
package me.shreyasr.chatse.chat.service

import android.util.Log
import okhttp3.Response
import okhttp3.ResponseBody
import okhttp3.ws.WebSocketListener
import okio.Buffer
import org.codehaus.jackson.JsonNode
import org.codehaus.jackson.map.ObjectMapper
import java.io.IOException

/**
 * WebSocketListener that listens to websocket
 */
class ChatWebSocketListener(private val site: String, private val listener: ServiceWebsocketListener) : WebSocketListener {
    private val mapper = ObjectMapper()

    override fun onOpen(webSocket: okhttp3.ws.WebSocket?, response: Response?) {
        Log.i("ChatWebSocketListener", "websocket open: " + site)
        listener.onConnect(site, true)
    }

    override fun onFailure(e: IOException?, response: Response?) {
        listener.onConnect(site, false)
    }

    @Throws(IOException::class)
    override fun onMessage(payload: ResponseBody) {
//        val message = payload.
//        payload.close()
        try {
            val root = mapper.readTree(payload.string())
            listener.onNewEvents(site, root)
        } catch (e: IOException) {
            Log.e("ChatWebSocketListener", e.message)
        }

    }

    override fun onPong(payload: Buffer) {
        Log.i("ChatWebSocketListener", "websocket pong: $site")
    }

    override fun onClose(code: Int, reason: String) {
        Log.i("ChatWebSocketListener", "websocket close: $site: $code, $reason")
    }

    interface ServiceWebsocketListener {
        fun onNewEvents(site: String, root: JsonNode)

        fun onConnect(site: String, success: Boolean)
    }
}
